File: app/wav_list.c
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "ff.h"

#include "wav_list.h"

#if _USE_LFN != 0
#error "LFN is not supported."
#endif

static char wav_list[APP_WAV_LIST_MAX_LEN][APP_WAV_LIST_MAX_NAME_LEN];
static char *wav_list_sorted[APP_WAV_LIST_MAX_LEN];
static uint16_t wav_list_len = 0U;

static bool ends_with(const char *str, const char *suffix) {
  size_t str_len = strlen(str);
  size_t suffix_len = strlen(suffix);
  if (str_len < suffix_len) {
    return false;
  }
  return strcmp(&str[str_len - suffix_len], suffix) == 0;
}

static int compare_str(const void *a, const void *b) {
  return strcmp(*(const char **)a, *(const char **)b);
}

bool app_wav_list_init(const char *path) {
  DIR dir;
  if (f_opendir(&dir, path) != FR_OK) {
    return false;
  }

  while (true) {
    if (wav_list_len >= APP_WAV_LIST_MAX_LEN) {
      break;
    }

    FILINFO file_info;
    if (f_readdir(&dir, &file_info) != FR_OK) {
      f_closedir(&dir);
      return false;
    }
    if (file_info.fname[0] == '\0') {
      break;
    }
    if ((file_info.fattrib & AM_DIR) != 0) {
      continue;
    }
    if (!ends_with(file_info.fname, ".WAV")) {
      continue;
    }
    memcpy(wav_list[wav_list_len], file_info.fname,
           sizeof(wav_list[wav_list_len]));
    wav_list_sorted[wav_list_len] = wav_list[wav_list_len];
    ++wav_list_len;
  }

  f_closedir(&dir);

  qsort(wav_list_sorted, wav_list_len, sizeof(wav_list_sorted[0]), compare_str);
  return true;
}

size_t app_wav_list_get_len(void) { return wav_list_len; }

const char *app_wav_list_get_name_by_index(size_t index) {
  if (index >= wav_list_len) {
    return NULL;
  }
  return wav_list_sorted[index];
}

File: src/WavList.zig
const std = @import("std");
const st = @import("stm32cubemx");

const Directory = @import("fatfs.zig").Directory;

comptime {
    if (st._USE_LFN != 0) @compileError("LFN is not supported.");
}

pub const Error = std.mem.Allocator.Error || Directory.Error;

allocator: std.mem.Allocator,
file_names: std.ArrayList([:0]const u8),

pub fn init(allocator: std.mem.Allocator, path: [:0]const u8) Error!@This() {
    var self: @This() = .{
        .allocator = allocator,
        .file_names = .empty,
    };
    errdefer self.deinit();

    var dir = try Directory.open(path);
    defer dir.close();

    while (true) {
        const file_info = try dir.read();
        const file_name = std.mem.sliceTo(&file_info.fname, 0);

        if (file_name.len == 0) break;
        if ((file_info.fattrib & st.AM_DIR) != 0) continue;
        if (!std.mem.endsWith(u8, file_name, ".WAV")) continue;

        const file_name_owned = try self.allocator.dupeZ(u8, file_name);
        errdefer self.allocator.free(file_name_owned);
        try self.file_names.append(self.allocator, file_name_owned);
    }

    std.mem.sort([:0]const u8, self.file_names.items, {}, lessThan);

    return self;
}

pub fn deinit(self: *@This()) void {
    for (self.file_names.items) |file_name| {
        self.allocator.free(file_name);
    }
    self.file_names.deinit(self.allocator);
}

pub fn names(self: *const @This()) []const [:0]const u8 {
    return self.file_names.items;
}

fn lessThan(_: void, a: [:0]const u8, b: [:0]const u8) bool {
    return std.mem.lessThan(u8, a, b);
}
